src/native/eventpipe/ds-process-protocol.c

```diff
@@ -386,6 +529,78 @@ process_protocol_helper_get_process_info (
 	ep_exit_error_handler ();
 }
 
+static
+bool
+process_protocol_helper_get_process_info_2 (
+	DiagnosticsIpcMessage *message,
+	DiagnosticsIpcStream *stream)
+{
+	EP_ASSERT (message != NULL);
+	EP_ASSERT (stream != NULL);
+
+	bool result = false;
+	ep_char16_t *command_line = NULL;
+	ep_char16_t *os_info = NULL;
+	ep_char16_t *arch_info = NULL;
+	ep_char16_t *managed_entrypoint_assembly_name = NULL;
+	ep_char16_t *clr_product_version = NULL;
+	DiagnosticsProcessInfo2Payload payload;
+	DiagnosticsProcessInfo2Payload *process_info_2_payload = NULL;
+
+	command_line = ep_rt_utf8_to_utf16_string (ep_rt_diagnostics_command_line_get (), -1);
+	ep_raise_error_if_nok (command_line != NULL);
+
+	os_info = ep_rt_utf8_to_utf16_string (ep_event_source_get_os_info (), -1);
+	ep_raise_error_if_nok (os_info != NULL);
+
+	arch_info = ep_rt_utf8_to_utf16_string (ep_event_source_get_arch_info (), -1);
+	ep_raise_error_if_nok (arch_info != NULL);
+
+	managed_entrypoint_assembly_name = ep_rt_utf8_to_utf16_string (ep_rt_entrypoint_assembly_name_get_utf8 (), -1);
+	ep_raise_error_if_nok (managed_entrypoint_assembly_name != NULL);
+
+	clr_product_version = ep_rt_utf8_to_utf16_string (ep_rt_runtime_version_get_utf8 (), -1);
+	ep_raise_error_if_nok (clr_product_version != NULL);
+
+	process_info_2_payload = ds_process_info_2_payload_init (
```

Comment:
Nit: this paradigm seems a little awkward - passing in a pointer to the struct and having it return the same pointer if it succeeds. I think it makes more sense if either you pass in a pointer and return a bool, or have the method allocate and return a pointer and then NULL means it failed.

Reply:
The init could take a heap or stack allocated struct and initialize it (also used by _alloc implementation to init the allocated struct), for heap allocations there is always a _alloc function, if the type supports heap allocations, you should always check against NULL for failures when calling init methods.

Reply:
As Johan said, this is a pattern for allowing us to use the same `init` method for both heap and stack allocated structures.

Reply:
I agree that it makes sense to pass in a pointer to a structure to initialize it, my point was that returning the same pointer you passed in to indicate success is counter intuitive.

If a method returns a pointer, that says to me that the method is either allocating something and is handing you back the pointer or has some cache internally and is handing out a pointer to it. NULL in that case means it failed in allocation or initialization, or the data is otherwise unavailable.

If I pass in a pointer to memory, it seems strange odd a method would hand back the same pointer I just passed in to indicate success. I don't need a pointer to it, I already own and manage the lifetime of the data. I would expect the method to return a bool or an HRESULT style error code to indicate success/failure.

Reply:
It was done this way to make the case of init/fini "simpler" when using a stack allocated object and close to the alloc/free pattern, so from that perspective it could be viewed as passing in a pointer to allocated memory and init will give back a full constructed object that is represented by returned pointer (or NULL on error), so if we just ignore that you need to supply allocated memory as first parameter, alloc/init will have same behaviour.

```
EventPipeProviderConfiguration config;
EventPipeProviderConfiguration *provider_config = ep_provider_config_init (&config, ...);
ep_raise_error_if_nok (provider_config != NULL);
....
ep_provider_config_fini (provider_config);
```

we could (in theory) use alloca for allocation, or use init directly with ep_rt_object_alloc (as done in _alloc implementations).

But I see your point as well and I don't believe it will be hard to adjust the pattern if we believe its an issue.